### ISSUE TITLE
Namespacing issue (reported on community channel)

### ISSUE THREAD

**Issue by BeeWarloc** [fusetools/fuselibs-public#559]
I don't know any details here, hopefully someone with more experience in this area are able to investigate this.

There's no simple reproduction case for this yet, but it _do_ appear to be a regression from 1.2.1.

> one more issue report here.. I'm creating a Class based on the <Alive.Page> and I get an error when the ux:Class is prefixed with the namespace that I've been using for other things in the app
`build\Local\Designer\cache\ux13\SOS.PageNewAlert.g.uno(4.46): E3111: SOS.Alive does not contain type or namespace 'Page'. Could you be missing a package reference?`
>
>
> [7:37] 
`<Alive.Page ux:Class="SOS.PageNewAlert">` -> error
> `<Alive.Page ux:Class="PageNewAlert">` -> all good :thinking_face:
>
>
> [7:38] 
> `E3111: SOS.Alive does not contain type or namespace 'CarouselPage'. Could you be missing a package reference?` this is also happening, but I'm not even trying to create a class from CarousselPage
> duckers [7:39 AM] 
> Can you provide a simple repro?
> 
>
> igorls [7:40 AM] 
> Well I was trying to do that, but I wasn't able to make it fail
https://fusecommunity.slack.com/archives/C0BJ8V4TV/p1507527418000151
> beewarloc [2:57 PM] 
> Is this a regression, ie. it worked on 1.2.1?
>
>
> igorls [3:01 PM] 
> I have to check
> ...
> I think you are right, the problem doesn't happen on 1.2.1

**Comment by mortoray** [comment]
Possibly they have a class named `SOS` somewhere. You can't have a symbol used for both a class and namespace. That might explain why the one line works and the other not.

**Comment by BeeWarloc** [comment]
But wouldn't that also be a problem on 1.2.1?

>I think you are right, the problem doesn't happen on 1.2.1

**Comment by mortoray** [comment]
It might not be as the Uno compiler does tend to get more strict over time. It's just a possibility we'd like to eliminate before looking further (a reproduction would be required in any case).

**Comment by BeeWarloc** [comment]
It could be then that this is not a blocker, I'll leave it to the fuselibs team to investigate this further and remove the "Release blocker" label if it's not.

**Comment by kusma** [comment]
The problem here seems to be this generated code:
```Uno
namespace SOS
{
    [Uno.Compiler.UxGenerated]
    public partial class PageNewAlert: Alive.Page
    {
        readonly Fuse.Navigation.Router router;
        [Uno.Compiler.UxGenerated]
        class Template1: Alive.CarouselPage
        {
```
...The problem is that the app contains an `SOS.Alive`-namespace, and symbol-resolution prefers these over matches closer to the root namespace. If we instead generated this

```Uno
namespace SOS
{
    [Uno.Compiler.UxGenerated]
    public partial class PageNewAlert: global::Alive.Page
    {
        readonly Fuse.Navigation.Router router;
        [Uno.Compiler.UxGenerated]
        class Template1: global::Alive.CarouselPage
        {
```
... we should be good. That is, we're lacking `global::` before the Alive-namespace.

**Comment by BeeWarloc** [comment]
Could using `global::` everywhere be a solution to this problem? Or would that ruin the readability of the generated code?

**Comment by kusma** [comment]
@BeeWarloc: That would be a good fix, I think. I have the patch for that already.

**Comment by kusma** [comment]
But looking through the changes of the UX-compiler, it doesn't look to me like this is a regression.

And I'm unable to build the example with the previous release, due to it using newer features.

**Comment by kusma** [comment]
But, thinking about this a bit more, it might actually be a regression because some Uno namespace-resolving issues was fixed AFAIK. That bug could have previously masked this bug. If so, this is probably a release-blocker after all...

**Comment by kusma** [comment]
No, confirmed with the reporter that this actually didn't work in 1.2 with this setup. So I'm pretty sure it's not a regression.

**Comment by BeeWarloc** [comment]
Ok, then I agree we shouldn't try to fix it now. Even though the mentioned fix seems solid to me, it's a bit late in the release cycle to introduce something like that.

**Comment by kusma** [comment]
Yep, I completely agree.

**Comment by kusma** [comment]
Here's a minimal reproduction-case:
Foo.uno:
```Uno
namespace Foo
{
	public class Bar
	{
	}
}
```
Bar.ux:
```UX
<Foo.Bar ux:Class="Bar.Foo.Baz" />
```
> Bar.Foo.Baz.g.uno(4.35): E3111: Bar.Foo does not contain type or namespace 'Bar'. Did you perhaps misspell 'Baz'? Could you be missing a package reference?

**Comment by kusma** [comment]
This seems like the right fix:
```patch
diff --git a/Source/UX/Uno.UX.Markup.CodeGeneration/CodeGenerator.cs b/Source/UX/Uno.UX.Markup.CodeGeneration/CodeGenerator.cs
index 4c9d87a610..1213fa4b82 100644
--- a/Source/UX/Uno.UX.Markup.CodeGeneration/CodeGenerator.cs
+++ b/Source/UX/Uno.UX.Markup.CodeGeneration/CodeGenerator.cs
@@ -187,14 +187,14 @@ void BeginClass(Scope scope)
             {
                 var t = scope.DocumentScope as UXIL.ClassNode;
                 Emit("[Uno.Compiler.UxGenerated]");
-                Emit("public partial class " + t.GeneratedClassName.Surname + ": " + t.BaseType.FullName);
+                Emit("public partial class " + t.GeneratedClassName.Surname + ": global::" + t.BaseType.FullName);
                 PushScope();
             }
             else if (scope.DocumentScope is UXIL.TemplateNode)
             {
                 var t = scope.DocumentScope as UXIL.TemplateNode;
                 Emit("[Uno.Compiler.UxGenerated]");
-                Emit("public partial class " + t.GeneratedClassName.Surname + ": Uno.UX.Template");
+                Emit("public partial class " + t.GeneratedClassName.Surname + ": global::Uno.UX.Template");
                 PushScope();
                 EmitParentScopeClosureConstructor(scope);
             }
```

...but in general, we should probably have the UX compiler *always* use fully qualified types. Which would probably change some APIs.

**Comment by mortoray** [comment]
Given how short and common names are in fuselibs I'm surprised this hasn't come up before. Definitely we should be using a `global::` prefix on names we've already resolved.  

I can't think of anything in fuselibs that would have a problem with this -- since they already work with actual types, thus nothing changes at runtime.

**Comment by kusma** [comment]
Just thought I'd point this out: I tried out the patch above, and it makes uno complain about `global::object`. So I guess we say that BaseType.FullName returns "object" in that case? That sounds a bit strange, as I'd expect it to return "Uno.Object" instead...

**Comment by mortoray** [comment]
You probably need to resolve all fundamental types to just the name. `float`, `int`, `object`, etc. All others should get the `global::` treatment.

**Comment by kusma** [comment]
Yeah, as long as 'float', 'int', 'object' etc is the real "full" typename. I thought the name was Uno.Float, Uno.Int, Uno.Object etc, and 'float', 'int', 'object' etc were keyword-aliases...

**Comment by mortend** [comment]
@kusma: You can use `dt.QualifiedName`, but you need to handle any generic type arguments
```
        static string GetGlobalName(this DataType dt)
        {
            var retval = "global::" + dt.QualifiedName;

            if (dt.GenericArguments.Length > 0)
            {
                var first = true;
                var sb = new StringBuilder(retval);
                sb.Append("<");

                foreach (var a in dt.GenericArguments)
                {
                    if (!first)
                        sb.Append(", ");

                    sb.Append(GetGlobalName(a));
                    first = false;
                }

                sb.Append(">");
                retval = sb.ToString();
            }

            return retval;
        }
```
